content/post/performance-metrics-histogram.md

```diff
@@ -0,0 +1,245 @@
+
+1. We aggregate the histograms from a stream of events.
+    1. Local metrics are aggregated at the Metrics Receiver within each Wallaroo worker, and a histogram is produced per step per metrics period (typically 2 seconds).
+2. Once a metrics collection period elapses, the collection for that period is sent out to an external monitoring system.
+3. The histograms for each step are stored in a list L, sorted by time.
+4. The list for each step is stored in a hashtable H, keyed by step.
+  Note that this relationship is invertible and depends on how we prefer to query things.
+5. To get the latency histogram of a time window W, we choose all the histograms for a step and perform a bin-wise addition (since the bin ranges are identical across the histograms, this is safe). The result is the bin-wise sum of the histograms in the window, which is another histogram of the same type. We call this histogram S.
+    ![fig 1: performance metrics histogram](/images/post/performance-metrics-histogram/fig01_performance_metrics_histogram.png)
+
+6. We transform the histogram S into a cumulative histogram S', where each bin now includes the weight of all of the bins to the left of it (in addition to its own weight). 
+    ![fig 2: performance metrics cumulative histogram](/images/post/performance-metrics-histogram/fig02_performance_metrics_cumulative_histogram.png)
+
+7. We normalize the cumulative histogram to create a cumulative distribution D.
+    ![fig 3: performance metrics cumulative distribution](/images/post/performance-metrics-histogram/fig03_performance_metrics_cumulative_distribution.png)
+
+8. To answer whether the 99th- and 99.9th-percentile latencies are below our target, we simply compare the value in the appropriate bins to the target value. In the illustrated example, the fraction below 1s is 99.8%, and the fraction below 1ms is 24.67%, so it fails the SLA.
+   
+
+### Cost Analysis
+
+- Events are binned in a histogram with pre-determined bin value ranges
+    - e.g. 0-1us, 1us-10us, 10us-100us, 100us-1ms, 1ms-10ms, 10ms-100ms, 100ms-1s, 1s-10s, >10s
+    - Such a histogram is constructed for each second.  
+        __**Space cost: O(|H|) for |H| bins**__  
+        __**Time cost: O(|E|) for |E| events**__ (but this is essentially "free" because it's streaming data.)  
+- The histograms for each second are stored in a list, L.  
+    __**Space cost: O(|L|*|H|)**__  
+- L is sorted by the time. The histograms arrive in order, so this is free.
+- We select the window w from L.  
+    __**Space cost: O(|w|*|H|)**__  
+    __**Time cost: O(|w|)**__ (Note that here |w| is the _number of histograms_, whereas for percentiles |W| is the _number of events_, which is much larger.)
+- We perform a bin-wise addition of the histograms. This means that we take the number of events in a bin and add them across all of the histograms in W, and do the same for each bin. This produces an aggregate histogram (which is just another histogram!).  
```

Comment:
Wouldn't it make sense to do the bin-wise addition as the histograms are being selected?

Reply:
This is what's being done. I don't think whether we gather all of the selected histograms before doing the addition or adding each histogram as it's selected makes a difference to the space/time cost, though.

Reply:
Makes sense.